From fb5e91f1be073f7943b17a7954a12b58a73d8dae Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Sat, 19 Jul 2014 04:23:34 -0400
Subject: [pci] checkpatch: Deprecate DEFINE_PCI_DEVICE_TABLE

Message-id: <20140719042334.19438.9218.stgit@gir.stowe>
Patchwork-id: 85970
O-Subject: [RHEL7.1 PATCH 08/81] PCI/checkpatch: Deprecate DEFINE_PCI_DEVICE_TABLE
Bugzilla: 1110895
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110895
Upstream Status: 92e112fdbb3cb55b43390426501a7efacd893b96

commit 92e112fdbb3cb55b43390426501a7efacd893b96
Author: Joe Perches <joe@perches.com>
Date:   Fri Dec 13 11:36:22 2013 -0700

    PCI/checkpatch: Deprecate DEFINE_PCI_DEVICE_TABLE

    Prefer use of the direct definition of struct pci_device_id instead of
    indirection via macro DEFINE_PCI_DEVICE_TABLE.

    Update the PCI documentation to deprecate DEFINE_PCI_DEVICE_TABLE.  Update
    checkpatch adding --fix option.

    Signed-off-by: Joe Perches <joe@perches.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
    Reviewed-by: Jingoo Han <jg1.han@samsung.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/Documentation/PCI/pci.txt b/Documentation/PCI/pci.txt
index 6f45856..9518006 100644
--- a/Documentation/PCI/pci.txt
+++ b/Documentation/PCI/pci.txt
@@ -123,8 +123,10 @@ initialization with a pointer to a structure describing the driver
 
 
 The ID table is an array of struct pci_device_id entries ending with an
-all-zero entry; use of the macro DEFINE_PCI_DEVICE_TABLE is the preferred
-method of declaring the table.  Each entry consists of:
+all-zero entry.  Definitions with static const are generally preferred.
+Use of the deprecated macro DEFINE_PCI_DEVICE_TABLE should be avoided.
+
+Each entry consists of:
 
  vendor,device Vendor and device ID to match (or PCI_ANY_ID)
 
diff --git a/include/linux/pci.h b/include/linux/pci.h
index c7a9881..1803254 100644
--- a/include/linux/pci.h
+++ b/include/linux/pci.h
@@ -694,8 +694,7 @@ struct pci_driver_rh {
  * DEFINE_PCI_DEVICE_TABLE - macro used to describe a pci device table
  * @_table: device table name
  *
- * This macro is used to create a struct pci_device_id array (a device table)
- * in a generic manner.
+ * This macro is deprecated and should not be used in new code.
  */
 #define DEFINE_PCI_DEVICE_TABLE(_table) \
  const struct pci_device_id _table[]
diff --git a/scripts/checkpatch.pl b/scripts/checkpatch.pl
index b954de5..d9ec7a9 100755
--- a/scripts/checkpatch.pl
+++ b/scripts/checkpatch.pl
@@ -2330,10 +2330,13 @@ sub process {
     $herecurr);
                }
 
-# check for declarations of struct pci_device_id
-  if ($line =~ /\bstruct\s+pci_device_id\s+\w+\s*\[\s*\]\s*\=\s*\{/) {
-   WARN("DEFINE_PCI_DEVICE_TABLE",
-        "Use DEFINE_PCI_DEVICE_TABLE for struct pci_device_id\n" . $herecurr);
+# check for uses of DEFINE_PCI_DEVICE_TABLE
+  if ($line =~ /\bDEFINE_PCI_DEVICE_TABLE\s*\(\s*(\w+)\s*\)\s*=/) {
+   if (WARN("DEFINE_PCI_DEVICE_TABLE",
+     "Prefer struct pci_device_id over deprecated DEFINE_PCI_DEVICE_TABLE\n" . $herecurr) &&
+       $fix) {
+    $fixed[$linenr - 1] =~ s/\b(?:static\s+|)DEFINE_PCI_DEVICE_TABLE\s*\(\s*(\w+)\s*\)\s*=\s*/static const struct pci_device_id $1\[\] = /;
+   }
   }
 
 # check for new typedefs, only function parameters and sparse annotations
-- 
1.7.1